# Practica de Merge de Datasets

In [1]:
# Importar librerias

import pandas as pd
import numpy as np
import seaborn as sns

sns.set_theme(style="whitegrid")

## 1.Cargar los datasets
Lo primero que debemos hacer es cargar los datasets. Esto lo podemos hacer a través de Drive o subiendo el archivo al colab

In [2]:
#Importamos la libreria de drive en caso de tener los datos en el drive y usar colab

#from google.colab import drive
#drive.mount('/gdrive')

In [ ]:
# Opción Google drive
# Usamos pandas para leer el csv en la ruta de nuestro drive
#df1 = pd.read_csv('/gdrive/MyDrive/Monitoria IA/Practicas/Merge/Personas1.csv',    #Se ingresa la ruta completa a la ubicación del archivo de interés. El MyDrive de la ruta es tu "Mi Unidad", desde ahí vas ingresando a cada carpeta que tengas hasta llegar a tu archivo.
 #                na_values='?')              #Como el archivo contiene algunos caracteres ?, con este comando pandas los interpreta como nan (not a number).

#df2 = pd.read_csv('/gdrive/MyDrive/Monitoria IA/Practicas/Merge/Personas2.csv',
#                 na_values='?')

### Subir archivo a Colab
Para subir el archivo a colab, damos click en el icono de carpeta que está en el menú del lado izquierdo y luego presionamos el icono del archivo con una flecha hacia arriba.
Esto nos abrirá unan ventana con el explorador de archivos, donde podremos importar el dataset.
Finalmente, solo se importa usando el nombre del csv y definiendo cuales valores interpretará pandas como nan.

In [3]:
# Opción GITHUB
df1 = pd.read_csv('https://raw.githubusercontent.com/armandoordonez/eda_couse/main/data/Personas1.csv',na_values='?')              #Como el archivo contiene algunos caracteres ?, con este comando pandas los interpreta como nan (not a number).
df2 = pd.read_csv('https://raw.githubusercontent.com/armandoordonez/eda_couse/main/data/Personas2.csv',na_values='?')

## 2. Verificamos los DF

In [4]:
df1.head()

first_name   last_name    cc
0     Berget   Proughten  1495
1     Scotti  Yegorovnin   311
2    Clayton       Reide   914
3     Toddie     Haddick   671
4     Brynna    Ferrucci   692

Podemos ver que el primero consta de una lista de personas que tienen un nombre y apellido, así como el número de cedula.

In [5]:
df2.head()

salary                city    cc
0   $5432.91  Sabon Gari-Nangere   611
1   $6581.31    San Rafael Abajo   561
2  $14405.46            Khvānsār    13
3  $27742.78              Cemara  1191
4  $88920.27          Diên Khánh   943

El segundo df consta del salario, la ciudad y el número de cedula de una persona, solo que en este caso desconocemos el nombre.

## 3. Problema del merge
Si quisieramos poder saber el nombre y apellido de la persona que está ganando ese salario y vive en esa ciudad, ¿Cómo lo haríamos?


Para problemas como este, en el que tenemos dos o más datasets que necesitamos "pegar" tomando como referencia sus columnas, utilizamos la función merge que nos brinda Pandas.

In [6]:
merged_df = pd.merge(df1, df2, on='cc')

Una posible solución es la linea de código que está arriba, en la cual tomamos ambos df y los pegamos tomando como referencia los números de cedula.

In [7]:
merged_df.head()

first_name last_name    cc     salary          city
0    Clayton     Reide   914  $79239.05     Wierzbica
1     Toddie   Haddick   671  $75752.12        Heping
2     Brynna  Ferrucci   692  $53230.11       Thorold
3     Jarred     Spore   692  $53230.11       Thorold
4    Atlanta   Farnill  1207   $2875.12  Duwakkandung

Como podemos observar, el resultado nos muestra que tenemos una nueva tabla con nombre, apellido, numero de cedula, salario y ciudad, logrando así unir los datos.
En este caso, el df resultante puede no tener el mismo tamaño que los df iniciales, esto sería si hay cedulas que no coinciden en ambos dataset.

## Tipos de Merge

Los principales tipos de merge se pueden separar en 3:
*   Inner: Es el tipo de merge por defecto, solo toma los datos que coincidan en su columna de referencia para ambos dataset. (En el ejemplo fue la cc, si en alguno de los dos dataset no estaba la misma cc, no se agrega)
*   Left: Agrega los datos que estén en ambos dataset, pero también agrega los datos del dataset que pongamos primero en nuestra función merge. (En el ejemplo, el df1)
*   Right: Agrega los datos que estén en ambos dataset, pero también agrega los datos del dataset que pongamos primero en nuestra función merge. (En el ejemplo, el df1)






Para cambiar el tipo de Merge debemos añadir una instrucción a la función llamada "how":


```
pd.merge(df1,df2,on='cc',how="left")
```



In [8]:
merged_df_left = pd.merge(df1,df2,on='cc', how='left')

In [ ]:
merged_df_left.head()

first_name   last_name    cc     salary       city
0     Berget   Proughten  1495        NaN        NaN
1     Scotti  Yegorovnin   311        NaN        NaN
2    Clayton       Reide   914  $79239.05  Wierzbica
3     Toddie     Haddick   671  $75752.12     Heping
4     Brynna    Ferrucci   692  $53230.11    Thorold

Se puede apreciar que, por ejemplo, para los dos primeros datos, falta el salario y la ciudad. Esto es porque al ser de tipo left, no encontró la cedula 1495 en el df2, por lo que no pudo emparejarle un salario o ciudad, lo que deja esos datos en NaN.

In [9]:
merged_df_right = pd.merge(df1,df2,on='cc', how='right')

In [ ]:
merged_df_right.head()

first_name last_name    cc     salary                city
0        NaN       NaN   611   $5432.91  Sabon Gari-Nangere
1        NaN       NaN   561   $6581.31    San Rafael Abajo
2       Abel    Toplin    13  $14405.46            Khvānsār
3    Drucill   Jaqueme  1191  $27742.78              Cemara
4        NaN       NaN   943  $88920.27          Diên Khánh

En el caso del right, pasa algo similar pero en caso contrario, puesto que no encontró para el primero caso la cedula 611 en el df1, por lo que deja el nombre y apellido en NaN, así también para otros casos.

## Nota:
También es posible cambiar las columnas que se toman de referencia al momento de hacer merge.


```
pd.merge(df1,df2,on=['cc','city'])
```
Podemos tener una lista de columnas y al momento de hacer el merge, si lo tenemos en su forma estándar (inner), solo traerá y emparejará los datos que tenga la misma cedula y ciudad en ambos datasets.
